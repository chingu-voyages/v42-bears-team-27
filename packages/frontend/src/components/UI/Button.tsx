/* eslint-disable react/jsx-props-no-spreading */
import type {
  HTMLAttributes,
  RefAttributes,
  ForwardRefExoticComponent,
} from 'react';
import { forwardRef } from 'react';
import type { ThemeUIStyleObject } from 'theme-ui';

export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
  sx?: ThemeUIStyleObject;
  /**
   * Size of the component
   * @default 'md'
   */
  size?: 'lg' | 'md' | 'sm';
  /**
   * Variant style of the component
   * @default 'filled'
   */
  variant?: 'filled' | 'outlined';
  /**
   * Add rounded edges for the component
   * @default true
   */
  rounded?: boolean;
  /**
   * Type of the button element
   * @default 'button'
   */
  type?: 'button' | 'reset' | 'submit';
}

export const Button: ForwardRefExoticComponent<
ButtonProps & RefAttributes<HTMLButtonElement>
> = forwardRef(
  (
    {
      children,
      size = 'md',
      variant = 'filled',
      rounded = true,
      type = 'button',
      sx,
      ...rest
    },
    ref,
  ) => {
    const styles: ThemeUIStyleObject = {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      borderRadius: rounded ? 50 : 0,
      ...(size === 'lg' && {
        columnGap: 10,
        py: 12,
        px: 117,
        fontSize: 24,
      }),
      ...(size === 'md' && {
        columnGap: 10,
        py: 3,
        px: 85,
        fontSize: 18,
      }),
      ...(size === 'sm' && {
        columnGap: 10,
        py: 3,
        px: 40,
        fontSize: 18,
      }),
    };

    return (
      <button
        // eslint-disable-next-line react/button-has-type
        type={type}
        ref={ref}
        sx={{ variant: `buttons.primary.${variant}`, ...styles, ...sx }}
        {...rest}
      >
        <p sx={{ variant: 'text.label', m: 0, fontSize: 'inherit' }}>
          {children}
        </p>
      </button>
    );
  },
);

if (process.env.NODE_ENV !== 'production') {
  Button.displayName = 'Button';
}
