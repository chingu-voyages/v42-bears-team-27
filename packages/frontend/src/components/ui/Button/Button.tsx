import type {
  HTMLAttributes,
  RefAttributes,
  ForwardRefExoticComponent,
} from 'react';
import { forwardRef } from 'react';
import Link, { type LinkProps } from 'next/link';
import type { ThemeUIStyleObject } from 'theme-ui';

export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
  /**
   * Type of the button element
   * @default 'button'
   */
  type?: 'button' | 'reset' | 'submit';
  /**
   * Size of the component
   * @default 'md'
   */
  size?: 'lg' | 'md' | 'sm';
  /**
   * Variant style of the component
   * @default 'filled'
   */
  variant?: 'filled' | 'outlined';
  /**
   * Add rounded edges for the component
   * @default true
   */
  rounded?: boolean;
  /**
   * Icon for the component
   */
  icon?: React.ReactNode;
  sx?: ThemeUIStyleObject;
}

export const Button: ForwardRefExoticComponent<
ButtonProps & RefAttributes<HTMLButtonElement>
> = forwardRef(
  (
    {
      children,
      icon,
      size = 'md',
      variant = 'filled',
      rounded = true,
      type = 'button',
      sx,
      ...rest
    },
    ref,
  ) => {
    const styles: ThemeUIStyleObject = {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      borderRadius: rounded ? 50 : 0,
      cursor: 'pointer',
      ...(size === 'lg' && {
        columnGap: 10,
        height: 54,
        py: 3,
        px: 5,
        fontSize: 24,
      }),
      ...(size === 'md' && {
        columnGap: 10,
        height: 48,
        py: 2,
        px: 4,
        fontSize: 18,
      }),
      ...(size === 'sm' && {
        columnGap: 10,
        height: 42,
        py: 2,
        px: 3,
        fontSize: 18,
      }),
    };

    return (
      <button
        // eslint-disable-next-line react/button-has-type
        type={type}
        ref={ref}
        sx={{ variant: `buttons.primary.${variant}`, ...styles, ...sx }}
        {...rest}
      >
        <p sx={{ variant: 'text.label', m: 0, fontSize: 'inherit' }}>
          {children}
        </p>
        {icon}
      </button>
    );
  },
);

if (process.env.NODE_ENV !== 'production') {
  Button.displayName = 'Button';
}

export interface ButtonLinkProps extends LinkProps {
  /**
   * Content of the component
   */
  children: React.ReactNode;
  /**
   * Type of the button element
   * @default 'button'
   */
  type?: 'button' | 'reset' | 'submit';
  /**
   * Size of the component
   * @default 'md'
   */
  size?: 'lg' | 'md' | 'sm';
  /**
   * Variant style of the component
   * @default 'filled'
   */
  variant?: 'filled' | 'outlined';
  /**
   * Add rounded edges for the component
   * @default true
   */
  rounded?: boolean;
  /**
   * Icon for the component
   */
  icon?: React.ReactNode;
  sx?: ThemeUIStyleObject;
}

export const ButtonLink: ForwardRefExoticComponent<
ButtonLinkProps & RefAttributes<HTMLAnchorElement>
> = forwardRef(
  (
    {
      children,
      icon,
      size = 'md',
      variant = 'filled',
      rounded = true,
      sx,
      ...rest
    },
    ref,
  ) => {
    const styles: ThemeUIStyleObject = {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      textDecoration: 'none',
      borderRadius: rounded ? 50 : 0,
      cursor: 'pointer',
      ...(size === 'lg' && {
        columnGap: 10,
        height: 54,
        py: 3,
        px: 5,
        fontSize: 24,
      }),
      ...(size === 'md' && {
        columnGap: 10,
        height: 48,
        py: 2,
        px: 4,
        fontSize: 18,
      }),
      ...(size === 'sm' && {
        columnGap: 10,
        height: 42,
        py: 2,
        px: 3,
        fontSize: 18,
      }),
    };

    return (
      <Link
        ref={ref}
        sx={{ variant: `buttons.primary.${variant}`, ...styles, ...sx }}
        {...rest}
      >
        <p sx={{ variant: 'text.label', m: 0, fontSize: 'inherit' }}>
          {children}
        </p>
        {icon}
      </Link>
    );
  },
);

if (process.env.NODE_ENV !== 'production') {
  ButtonLink.displayName = 'ButtonLink';
}
