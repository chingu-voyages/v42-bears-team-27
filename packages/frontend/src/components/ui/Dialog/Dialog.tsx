import type { RefAttributes, ForwardRefExoticComponent } from 'react';
import { forwardRef } from 'react';
import type { ThemeUIStyleObject } from 'theme-ui';
import * as DialogPrimitive from '@radix-ui/react-dialog';
import { MdClose } from 'react-icons/md';

import { IconButton } from '../IconButton';

export interface DialogContentProps extends DialogPrimitive.DialogContentProps {
  /**
   * Title of the component
   */
  title: string;
  /**
   * Description of the component
   */
  description?: string;
  /**
   * Trigger button to reveal component
   */
  /**
   * Width of the component
   */
  width: string | number;
  /**
   * Height of the component
   */
  height: string | number;
  sx?: ThemeUIStyleObject;
}

const overlayStyles: ThemeUIStyleObject = {
  position: 'fixed',
  top: 0,
  left: 0,
  zIndex: 1000,
  width: '100%',
  height: '100%',
  backdropFilter: 'blur(3px)',
  bg: 'rgba(0, 0, 0, 0.6)',
};

const contentStyles: ThemeUIStyleObject = {
  position: 'fixed',
  top: '50%',
  left: '50%',
  zIndex: 1001,
  maxWidth: 880,
  overflowY: 'auto',
  py: 2,
  px: 4,
  bg: 'muted',
  translate: '-50% -50%',
  transition: 'opacity 700ms',
};

export const DialogContent: ForwardRefExoticComponent<
DialogContentProps & RefAttributes<HTMLDivElement>
> = forwardRef(
  ({ children, title, description, width, height, sx, ...rest }, ref) => (
    <DialogPrimitive.Portal>
      <DialogPrimitive.Overlay sx={overlayStyles} />
      <DialogPrimitive.Content
        ref={ref}
        sx={{ ...contentStyles, width, height, ...sx }}
        {...rest}
      >
        <DialogPrimitive.Title sx={{ variant: 'text.h3', textAlign: 'center' }}>
          {title}
        </DialogPrimitive.Title>
        <DialogPrimitive.Description sx={{ variant: 'text.h4' }}>
          {description}
        </DialogPrimitive.Description>
        {children}
        <DialogPrimitive.Close asChild>
          <IconButton sx={{ position: 'absolute', top: 4, right: 4 }}>
            <MdClose size={32} />
          </IconButton>
        </DialogPrimitive.Close>
      </DialogPrimitive.Content>
    </DialogPrimitive.Portal>
  ),
);

if (process.env.NODE_ENV !== 'production') {
  DialogContent.displayName = 'DialogContent';
}

export const Dialog = DialogPrimitive.Root;
export const DialogTrigger = DialogPrimitive.Trigger;
