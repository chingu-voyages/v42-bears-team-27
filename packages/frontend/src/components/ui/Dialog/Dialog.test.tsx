import { render, screen } from 'test-utils';
import userEvent from '@testing-library/user-event';

import { Dialog, DialogContent, DialogTrigger } from './Dialog';
import { Button } from '../Button';

describe('<Dialog />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <Dialog>
          <DialogTrigger asChild>
            <Button>Foo</Button>
          </DialogTrigger>
        </Dialog>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Dialog state', () => {
    it('should display Dialog children when state open', async () => {
      render(
        <Dialog open>
          <DialogContent title="Foo" width={320} height={480}>
            Hello, World
          </DialogContent>
        </Dialog>,
      );
      await userEvent.click(screen.getByRole('button'));
      expect(screen.getByText(/hello, world/i));
    });
  });
});
