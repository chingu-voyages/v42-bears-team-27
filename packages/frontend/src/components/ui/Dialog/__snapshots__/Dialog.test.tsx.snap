// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog /> Snapshots should match snapshot 1`] = `
.emotion-0 {
  background-color: var(--theme-ui-colors-primary);
  color: var(--theme-ui-colors-text);
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-radius: 50px;
  -webkit-column-gap: 10px;
  column-gap: 10px;
  height: 48px;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 32px;
  padding-right: 32px;
  font-size: 18px;
}

.emotion-0:hover {
  background-color: var(--theme-ui-colors-primaryShade);
}

.emotion-1 {
  color: var(--theme-ui-colors-text);
  font-family: "Nunito",sans-serif;
  font-size: inherit;
  font-weight: 400;
  line-height: 1.5;
  margin: 0;
}

<div>
  <button
    aria-controls="radix-:r0:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="emotion-0"
    data-state="closed"
    type="button"
  >
    <p
      class="emotion-1"
    >
      Foo
    </p>
  </button>
</div>
`;
